cloud/pkg/controller/manager/location.go

```diff
@@ -4,18 +4,23 @@ import (
 	"fmt"
 	"sync"
 
-	"github.com/kubeedge/beehive/pkg/common/log"
 	"k8s.io/api/core/v1"
 )
 
 // LocationCache cache the map of node, pod, configmap, secret
 type LocationCache struct {
-	//edgeNodes is a list of valid edge nodes
-	edgeNodes []string
+	// EdgeNodes is a map, key is nodeName, value is Status
+	EdgeNodes map[string]string
+	// EdgeNodesLock is lock for edgeNodes
+	EdgeNodesLock sync.RWMutex
 	// configMapNode is a map, key is namespace/configMapName, value is nodeName
 	configMapNode sync.Map
 	// secretNode is a map, key is namespace/secretName, value is nodeName
 	secretNode sync.Map
+	// Services is an array of services
+	Services []v1.Service
```

Comment:
1. if we use []string, time complexity of function UpdateService and DeleteService is O(n), but sync service to edge when nodestate from ```not ready``` to ```ready``` is O(1)

2. if we use sync.Map, time complexity of function UpdateService and DeleteService is O(1), but sync service to edge when nodestate from ```not ready``` to ```ready``` is O(n)

Usually, service count is more than node count. service update event is more than node state update event.
on the other hand, add or update service in map is clearer than in []string

So i think, method 2 is better than method 1, can be considered, no mandatory

*note: sync.Map(key: namespace/serviceName, value: v1.Service)*

Reply:
Thanks for pointing it out. Used sync.Map